### ISSUE TITLE
new rule - detect unnecessary instance/class methods

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#6582]
```py
class Foo:
    def foo(self): # unnecessary instance method, the method does not access the `self` argument
        pass

    @classmethod
    def bar(cls): # unnecessary class method, the method does not access the `cls` argument
        pass
```

the rule could suggest making them static methods or functions outside of the class

**Comment by zanieb** [comment]
This sounds good to me — there aren't any prior implementations of this?

**Comment by charliermarsh** [comment]
My only concern is that we may get a lot of false positives for methods that are intended to be overridden by subclasses but aren't abstract, but I haven't thought about it deeply.

**Comment by zanieb** [comment]
We could add an exclusion for methods that raise a `NotImplementedError` or have an empty body e.g. `pass` or `...`?

**Comment by charliermarsh** [comment]
Yeah, I'd be interested to see the false positive rate. I realized that we actually have the instance method version open as https://github.com/astral-sh/ruff/pull/6574 (recent PR).

Separately, I was curious, so I confirmed that Mypy does not allow this:

```python
class Foo:
    @staticmethod
    def foo() -> None:
        pass

class Bar(Foo):
    def foo(self) -> None:
        print(1)

Bar().foo()
```

**Comment by DetachHead** [comment]
lol i had no idea this was covered by #970 because [`no-self-use` doesn't work unless you enable its plugin](https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/no-self-use.html)

**Comment by jonathandung** [comment]
This can be closed.

**Comment by MichaReiser** [comment]
Thanks for the ping. This is implemented as https://docs.astral.sh/ruff/rules/no-self-use/ (in preview)
